src/parser2.rs

```diff
@@ -0,0 +1,310 @@
+use pest::prelude::*;
+
+
+impl_rdp! {
+    grammar! {
+        // basic blocks of the language
+        op_or        = { ["or"] }
+        op_wrong_or  = { ["||"] }
+        op_and       = { ["and"] }
+        op_wrong_and = { ["&&"] }
+        op_lte       = { ["<="] }
+        op_gte       = { [">="] }
+        op_lt        = { ["<"] }
+        op_gt        = { [">"] }
+        op_eq        = { ["=="] }
+        op_ineq      = { ["!="] }
+        op_plus      = { ["+"] }
+        op_minus     = { ["-"] }
+        op_times     = { ["*"] }
+        op_slash     = { ["/"] }
+        op_true      = { ["true"] }
+        op_false     = { ["false"] }
+        boolean      = _{ op_true | op_false }
+
+        int   = @{ ["-"]? ~ (["0"] | ['1'..'9'] ~ ['0'..'9']*) }
+        float = @{
+            ["-"]? ~
+                ["0"] ~ ["."] ~ ['0'..'9']+ |
+                ['1'..'9'] ~ ['0'..'9']* ~ ["."] ~ ['0'..'9']+
+        }
+
+        identifier = @{
+            (['a'..'z'] | ['A'..'Z'] | ["_"]) ~
+            (['a'..'z'] | ['A'..'Z'] | ["_"] | ["."] | ['0'..'9'])*
+        }
+        // matches anything between 2 double quotes
+        string = @{ ["\""] ~ (!(["\""]) ~ any )* ~ ["\""]}
+
+        // Precedence climbing
+        expression = _{
+            // boolean first so they are not caught as identifiers
+            { boolean | identifier | int | float }
+            or          = { op_or | op_wrong_or }
+            and         = { op_and | op_wrong_and }
+            comparison  = { op_gt | op_lt | op_eq | op_ineq | op_lte | op_gte }
+            add_sub     = { op_plus | op_minus }
+            mul_div     = { op_times | op_slash }
+        }
+
+        // Tera specific things
+
+        // different types of blocks
+        variable_start = _{ ["{{"] }
+        variable_end   = _{ ["}}"] }
+        tag_start      = _{ ["{%"] }
+        tag_end        = _{ ["%}"] }
+        comment_start  = _{ ["{#"] }
+        comment_end    = _{ ["#}"] }
+        block_start    = { variable_start | tag_start | comment_start }
+
+        // Actual tags
+        extends_tag  = { tag_start ~ ["extends"] ~ string ~ tag_end }
+
+        variable_tag = { variable_start ~ expression ~ variable_end }
+        comment_tag  = { comment_start ~ (!comment_end ~ any )* ~ comment_end }
+        block_tag    = { tag_start ~ ["block"] ~ identifier ~ tag_end }
+        if_tag       = { tag_start ~ ["if"] ~ expression ~ tag_end }
+        elif_tag     = { tag_start ~ ["elif"] ~ expression ~ tag_end }
+        else_tag     = { tag_start ~ ["else"] ~ tag_end }
+        for_tag      = { tag_start ~ ["for"] ~ identifier ~ ["in"] ~ identifier ~ tag_end }
+        endblock_tag = { tag_start ~ ["endblock"] ~ identifier ~ tag_end }
+        endif_tag    = { tag_start ~ ["endif"] ~ tag_end }
+        endfor_tag   = { tag_start ~ ["endfor"] ~ tag_end }
+
+        text = { (!(block_start) ~ any )* }
```

Comment:
`end()` fails because `whitespace` gets run only in front of any rule. (including `eoi`) This example has some white-space at the end which is why `end()` returns `false`.